Assets/LeapMotion/Core/Scripts/XR/XRHeightOffset.cs

```diff
@@ -0,0 +1,129 @@
+
+using UnityEngine;
+using System;
+using System.Linq;
+using Leap.Unity.Attributes;
+
+public class XRHeightOffset : MonoBehaviour {
+
+  #region Inspector
+
+  [Header("Height Offset When XR Space Tracking Type Is Stationary")]
+
+  [SerializeField]
+  [OnEditorChange("stationaryHeightOffset")]
+  [Tooltip("This offset only used when the XR space tracking type is 'Stationary.' This "
+         + "is the tracking type for Oculus setups with a single tracking camera and "
+         + "for Android VR devices without position tracking. Two-camera Oculus setups "
+         + "and Vive headsets, for example, do not require a height offset. They center "
+         + "the room-scale floor on the Rig transform, and this script will not apply "
+         + "a height offset when the RoomScale space-tracking type is detected.")]
+  private float _stationaryHeightOffset = 1.6f; // average human height (or so)
+  public float stationaryHeightOffset {
+    get { return _stationaryHeightOffset; }
+    set {
+      _stationaryHeightOffset = value;
+
+      applyHeightOffset();
+    }
+  }
+
+  #region Auto Recenter
+
+  [Header("Auto Recenter")]
+
+  [SerializeField]
+  [Tooltip("If the detected XR device is present and supports userPresence, "
+         + "checking this option will detect when the user puts on the device headset "
+         + "and call InputTracking.Recenter.")]
+  private bool autoRecenterOnUserPresence = true;
+
+  private UnityEngine.XR.UserPresenceState _lastUserPresence;
+
+  #endregion
+
+  #region Runtime Height Adjustment
+
+  [Header("Runtime Height Adjustment")]
+
+  public bool enableRuntimeAdjustment = true;
+
+  [DisableIf("enableRuntimeAdjustment", isEqualTo: false)]
+  [Tooltip("Press this key on the keyboard to adjust the height offset up by stepSize.")]
+  public string stepUpKey = "up";
+
+  [DisableIf("enableRuntimeAdjustment", isEqualTo: false)]
+  [Tooltip("Press this key on the keyboard to adjust the height offset down by stepSize.")]
+  public string stepDownKey = "down";
+
+  [DisableIf("enableRuntimeAdjustment", isEqualTo: false)]
+  public float stepSize = 0.1f;
+
+  #endregion
+
+  #endregion
+
+  #region Unity Events
+
+  private Vector3 _startingPosition = Vector3.zero;
+
+  private void Start() {
+    _startingPosition = this.transform.position;
+    applyHeightOffset();
+
+    // Auto recenter
+    var userPresence = UnityEngine.XR.XRDevice.userPresence;
+
+    if (userPresence == UnityEngine.XR.UserPresenceState.Unsupported) {
+      Debug.Log("[XRAutoRecenter] XR UserPresenceState unsupported; "
+              + "disabling autoRecenterOnUserPresence. (XR support is probably disabled.)");
+      autoRecenterOnUserPresence = false;
+    }
+  }
+
+  private void Update() {
+    var deviceIsPresent = UnityEngine.XR.XRDevice.isPresent;
+    if (deviceIsPresent) {
+      
+      if (enableRuntimeAdjustment) {
+        if (Input.GetKeyDown(stepUpKey)) {
```

Comment:
I would favor using actual KeyCodes over using named input fields, since we don't know what the users input state might be like.

Reply:
KeyCode enums just have a terrible inspector experience.. :(

Reply:
I would argue strings are even worse :P You have to type in text and you don't know if you are correct or incorrect until you press play.

Reply:
When I made the change I had forgotten that the strings correspond to InputManager entries; if you could type a string and have it match to a KeyCode of the same name, that's nice and quick assuming you know the name of the key you want.

Now it's on my list to add an attribute for KeyCodes that will improve its inspector -- I think a dropdown with a properly categorized menu would be a big improvement